pkg/queue/concurrency_state.go

```diff
@@ -38,50 +40,48 @@ func ConcurrencyStateHandler(logger *zap.SugaredLogger, h http.Handler, pause, r
 		resume = func() {}
 	}
 
-	type req struct {
-		w http.ResponseWriter
-		r *http.Request
+	var (
+		inFlight = atomic.NewInt64(0)
+		paused   = true
+		mux      sync.RWMutex
+	)
 
-		done chan struct{}
-	}
-
-	reqCh := make(chan req)
-	doneCh := make(chan struct{})
-	go func() {
-		inFlight := 0
-
-		// This loop is entirely synchronous, so there's no cleverness needed in
-		// ensuring open and close dont run at the same time etc. Only the
-		// delegated ServeHTTP is done in a goroutine.
-		for {
-			select {
-			case <-doneCh:
-				inFlight--
-				if inFlight == 0 {
-					logger.Info("Requests dropped to zero ...")
+	return func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
```

Comment:
if add mux.Lock here, not line 52, need double check again?

Reply:
You don't want to fetch the Lock in all cases, that'll make this quite a lot slower again. The atomic integer is very fast even under a lot of contention. A write lock is slow in comparison so we delay getting it to only the (comparatively rare) edge-cases where it's really necessary.